### PR TITLE
V3.6.0 pipeline

### PR DESCRIPTION
**Pull request by star-e**
Refactoring EffectAsset, ProgramLib, (Material) Pass.

In custom pipeline, we will watch all effectAssets. If any effectAsset has been loaded/unloaded, we will rebuild descriptor layout graph before rendering.
This implies:
1. Game might stuck in the first render frame when all effects are loaded.
2. material and (material) pass require rebuilding.

### Changelog

* Modified effect-asset, program-lib, root.
* Updated custom pipeline.

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.

  > Manual trigger with `@cocos-robot run test cases` afterward.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

**Comment by star-e** [comment]
@cocos-robot run test cases

File: cocos/core/assets/effect-asset.ts [hunk 1]
```diff
@@ -212,7 +175,10 @@ export class EffectAsset extends Asset {
      * @en Register the effect asset to the static map
      * @zh 将指定 effect 注册到全局管理器。
      */
-    public static register (asset: EffectAsset) { EffectAsset._effects[asset.name] = asset; }
+    public static register (asset: EffectAsset) {
+        EffectAsset._effects[asset.name] = asset;
+        EffectAsset._layoutValid = false;
+    }
```
Comment (star-e) [inline cocos/core/assets/effect-asset.ts:182]: refactoring shader descriptors.

File: cocos/core/renderer/core/program-lib.ts [hunk 2]
```diff
@@ -442,9 +443,28 @@ class ProgramLib {
...
+                if (globalDS) {
+                    tmplInfo.setLayouts[SetIndex.GLOBAL] = globalDS;
+                }
+                if (materialDS) {
+                    tmplInfo.setLayouts[SetIndex.MATERIAL] = materialDS;
+                }
+                if (localDS) {
+                    tmplInfo.setLayouts[SetIndex.LOCAL] = localDS;
+                }
+            } else {
```
Comment (star-e) [inline cocos/core/renderer/core/program-lib.ts:463]: Adding custom pipeline, refactoring ProgramLib

Comment (minggo) [reply cocos/core/renderer/core/program-lib.ts:463]: Should also modify native implementation?

Comment (star-e) [reply cocos/core/renderer/core/program-lib.ts:463]: Yes, native implementation will be added later. Currently custom pipeline cannot be enabled yet, so this code path will not be reached yet.

Comment (minggo) [reply cocos/core/renderer/core/program-lib.ts:463]: Got it.

**Review by minggo — APPROVED** [review]

**Pull request merged**